src/k2/module/k23si/Module.cpp

```diff
@@ -159,24 +163,192 @@ _makeReadOK(dto::DataRecord* rec) {
     return RPCResponse(dto::K23SIStatus::OK("read succeeded"), std::move(response));
 }
 
-seastar::future<std::tuple<Status, dto::K23SIReadResponse>>
-K23SIPartitionModule::handleRead(dto::K23SIReadRequest&& request, dto::K23SI_MTR sitMTR, FastDeadline deadline) {
-    K2DEBUG("Partition: " << _partition << ", received read " << request);
-    if (!_validateRequestPartition(request)) {
-        // tell client their collection partition is gone
-        return RPCResponse(dto::K23SIStatus::RefreshCollection("collection refresh needed in read"), dto::K23SIReadResponse{});
+// Helper for iterating over the indexer, returns end() if iterator would go past the target schema
+// or if it would go past begin() for reverse scan. Starting iterator must not be end() and must 
+// point to target record
+void scanAdvance(std::map<dto::Key, std::deque<dto::DataRecord>>::iterator& it, 
+            std::map<dto::Key, std::deque<dto::DataRecord>>& map, bool reverseDirection) {
+    const String& schema = it->first.schemaName;
+
+    if (!reverseDirection) {
+        ++it;
+        if (it != map.end() && it->first.schemaName != schema) {
+            it = map.end();
+        }
+
+        return;
     }
-    if (!_validateRequestParameter(request)){
-        // do not allow empty partition key
-        return RPCResponse(dto::K23SIStatus::BadParameter("missing partition key in read"), dto::K23SIReadResponse{});
+
+    if (it == map.begin()) {
+        it = map.end();
+    } else {
+        --it;
+
+        if (it->first.schemaName != schema) {
+            it = map.end();
+        }
     }
-    if (!_validateRetentionWindow(request)) {
-        // the request is outside the retention window
-        return RPCResponse(dto::K23SIStatus::AbortRequestTooOld("request too old in read"), dto::K23SIReadResponse{});
+}
+
+seastar::future<std::tuple<Status, dto::K23SIQueryResponse>>
+K23SIPartitionModule::handleQuery(dto::K23SIQueryRequest&& request, dto::K23SIQueryResponse&& response, FastDeadline deadline) {
+    K2DEBUG("Partition: " << _partition << ", received query " << request);
+
+    Status validateStatus = _validateReadRequest(request);
+    if (!validateStatus.is2xxOK()) {
+        return RPCResponse(std::move(validateStatus), dto::K23SIQueryResponse{});
     }
-    if (_schemas.find(request.key.schemaName) == _schemas.end()) {
-        // server does not have schema
-        return RPCResponse(dto::K23SIStatus::OperationNotAllowed("schema does not exist"), dto::K23SIReadResponse{});
+    if (_partition.getHashScheme() != dto::HashScheme::Range) {
+            return RPCResponse(dto::K23SIStatus::OperationNotAllowed("Query is only allowed on range partitioned collection"), dto::K23SIQueryResponse{});
```

Comment:
yeah, we can do it for hash partitioning but it will require some implementation on the client side. I changed to message for now to indicate it isn't implemented yet.